app/backend/src/couchers/interceptors.py

```diff
@@ -34,26 +70,43 @@ class AuthValidatorInterceptor(grpc.ServerInterceptor):
     terminates the call with an UNAUTHENTICATED error code.
     """
 
-    def __init__(self, get_session_for_token, allow_jailed=True):
-        self._get_session_for_token = get_session_for_token
-        self._allow_jailed = allow_jailed
+    def __init__(self):
+        self._pool = get_descriptor_pool()
 
     def intercept_service(self, continuation, handler_call_details):
-        token = parse_session_cookie(dict(handler_call_details.invocation_metadata))
+        method = handler_call_details.method
+        # method is of the form "/org.couchers.api.core.API/GetUser"
+        _, service_name, method_name = method.split("/")
 
-        if not token:
-            return unauthenticated_handler()
+        service_options = self._pool.FindServiceByName(service_name).GetOptions()
+        auth_level = service_options.Extensions[annotations_pb2.auth_level]
 
-        # None or (user_id, jailed)
-        res = self._get_session_for_token(token=token)
+        # if unknown auth level, then it wasn't set and something's wrong
+        if auth_level == annotations_pb2.AUTH_LEVEL_UNKNOWN:
+            raise Exception("Service not annotated with auth_level")
 
-        if not res:
-            return unauthenticated_handler()
+        assert auth_level in [
+            annotations_pb2.AUTH_LEVEL_OPEN,
+            annotations_pb2.AUTH_LEVEL_JAILED,
+            annotations_pb2.AUTH_LEVEL_SECURE,
+        ]
+
+        token = parse_session_cookie(dict(handler_call_details.invocation_metadata))
+
+        res = _try_get_and_update_user_details(token)
 
-        user_id, jailed = res
+        # if no session was found and this isn't an open service, fail
+        if not token or not res:
+            if auth_level != annotations_pb2.AUTH_LEVEL_OPEN:
+                return unauthenticated_handler()
+            user_id = None
+        else:
+            # a valid user session was found
+            user_id, is_jailed = res
 
-        if not self._allow_jailed and jailed:
-            return unauthenticated_handler("Permission denied")
+            # if the user is jailed and this is isn't a jailed service, fail
+            if is_jailed and auth_level != annotations_pb2.AUTH_LEVEL_JAILED:
```

Comment:
Jailed users should be able to call the open service right?

Reply:
Yes... you've uncovered a heisenbug before it entered prod